### ISSUE TITLE
dataframe.describe() takes lot of time to load memory object into RAM

### ISSUE THREAD

**Issue by Rakesh-E532238** [vaexio/vaex#801]
HI,

Thanks for giving vaex. I have just started working on it and I am really excited!

I have an issue after I merge all my hdf5 files into a hd5 file and load it. While I do a describe on the dataframe it takes me about 40 mins for a 200mn-250mn data. And it happens everytime I restart the kernel. I use Dell 2.70GHz Intel Core i7, 32GB RAM (WINDOWS) machine with 8 logical processors.

How do I handle this more efficiently? Can I store the memory mapped object and use it right away instead of copying into RAM?

**Comment by maartenbreddels** [comment]
Hi,

thanks for the kinds words! That is strange, 50gb/(40*60sec)=20mb/s, which is too slow for any harddrive, do you have a spinner hdd, or and SSD/nvme drive?
How did you add the files together?

cheers,

Maarten

**Comment by Rakesh-E532238** [comment]
> Hi,
> 
> thanks for the kinds words! That is strange, 50gb/(40*60sec)=20mb/s, which is too slow for any harddrive, do you have a spinner hdd, or and SSD/nvme drive?
> How did you add the files together?
> 
> cheers,
> 
> Maarten

It's an HDD.

**Comment by Rakesh-E532238** [comment]
> > Hi,
> > thanks for the kinds words! That is strange, 50gb/(40*60sec)=20mb/s, which is too slow for any harddrive, do you have a spinner hdd, or and SSD/nvme drive?
> > How did you add the files together?
> > cheers,
> > Maarten
> 
> It's an HDD.

I used the below code for adding and exporting to one hd5 file.

for filename in os.listdir('dir_csv_path'):
    try:
        pandas_df= pd.read_csv(dir_csv_path+filename, encoding="utf-8")
        vaex_df = vaex.from_pandas(pandas_df, copy_index=False)
        vaex_df.export_hdf5(path=output+filename.split(".")[0]+"hdf5", progress=False)
        print(dir_csv_path + filename)
    except Exception as e:
        print("filename Failed: "+filename)

import re
import glob
import vaex
import numpy as np

def tryint(s):
    try:
        return int(s)
    except:
        return s

def alphanum_key(s):
    """ Turn a string into a list of string and number chunks.
        "z23a" -> ["z", 23, "a"]
    """
    return [ tryint(c) for c in re.split('([0-9]+)', s) ]

import os
os.chdir('../EA/data/')
hdf5_list = glob.glob('*.hdf5')
print(len(hdf5_list))
hdf5_list.sort(key=alphanum_key)
hdf5_list = np.array(hdf5_list)

master_df = vaex.open_many(hdf5_list)

# exporting
master_df.export_hdf5(path='../mapped_data/data_2016_2019.hd5', progress=True)

**Comment by maartenbreddels** [comment]
(I reformatted your issue with triple backticks to make it easier to read).

This looks good. To be honest, I don't think you should try working with a HDD with so much data. A modern nvme drive can read 4GB/s. But still, the 20mb/s doesn't seem right. Maybe if your HDD is almost full it's slow, but this slow is strange.

**Comment by Rakesh-E532238** [comment]
Sure. I still have 400GB data available for use.

**Comment by maartenbreddels** [comment]
> And it happens everytime I restart the kernel.

What do you mean by this btw, what happens if you don't restart the kernel, can you elaborate?

**Comment by Rakesh-E532238** [comment]
When I restart the kernel the RAM copies the mapping data again at a slow rate. It happened for the first 4-5 attempts. 

I just tested now, it took 3-4 mins now show the results of describe() function. I am not sure how the improvement happened now and why it took 5 attempts to read a faster rate.

**Comment by Rakesh-E532238** [comment]
I sometimes restart the kernel to let go of the memory occupied in RAM(which is usually around 85-95%) to perform other application activities.

**Comment by maartenbreddels** [comment]
So, you only saw the slow reading the first (few?) times. But now it is consistently faster (3-4 minutes)?

**Comment by Rakesh-E532238** [comment]
yes.

**Comment by Rakesh-E532238** [comment]
May I know how the memory map object works for this huge file. What gets copied into the RAM when we calculate stats on the dataframe? Is it(memory map) binded with the hd5 file while we export or is it something creating when we first call describe()?

**Comment by Rakesh-E532238** [comment]
It has come down to 30-40s now. I will look into disk copying aspect. I will reach out to you if I face an issue again. 

I also wanted to understand if we could do the same merging of files (pasted above) like hd5 using parquet files. Is there any other alternate using parquet and still achieve the same performance boost?

Thanks for your prompt responses. Appreciate it.

**Comment by maartenbreddels** [comment]
There is support for Arrow, but until that improves I would stay with hdf5 for now. #517 would require some more work. There is not direct funding on that ATM, so it is not the highest priority, but it will most likely happen before the end of the year.

**Comment by Rakesh-E532238** [comment]
Hi Maarten,

Does having multiple hd5 files from a directory instead of one big hd5 files have an impact on how vaex performs? 

I am about to generate two files (one 640GB and another 320 GB). I will be merging them into one (960GB) later. The configuration I am using is

-  128GB RAM, 2 Dual xeon processors
-  HDD (Read/Write speed of around around 170-190MB/s). Will soon plan to move to SSD.

Any suggestions?

**Comment by maartenbreddels** [comment]
> I am about to generate two files (one 640GB and another 320 GB). I will be merging them into one (960GB) later.

Now, that's a real use case for vaex :)



> Does having multiple hd5 files from a directory instead of one big hd5 files have an impact on how vaex performs?

We've noticed it matters a bit sometimes, sometimes quite a bit. I couldn't really understand why. Maybe you can let us know if you notice it!



> Any suggestions?

Try and let us know please :)

**Comment by Rakesh-E532238** [comment]
> > I am about to generate two files (one 640GB and another 320 GB). I will be merging them into one (960GB) later.
> 
> Now, that's a real use case for vaex :)
> 
> > Does having multiple hd5 files from a directory instead of one big hd5 files have an impact on how vaex performs?
> 
> We've noticed it matters a bit sometimes, sometimes quite a bit. I couldn't really understand why. Maybe you can let us know if you notice it!
> 
> > Any suggestions?
> 
> Try and let us know please :)

sure :)

**Comment by maartenbreddels** [comment]
My guess is the performance difference is small, when you have 1000 files
it will probably be more noticeable.

(from mobile phone)

**Comment by maartenbreddels** [comment]
> * Will soon plan to move to SSD.

small tip, 1 nvme drive can read at 4GB/s, but 2 in raid0 can read at 8GB/s, at the same price (two drives at half capacity), 4 will go at 16GB/s, etc. I've seen the best performance on Linux (at least that's what @mariobuikhuizen showed me, windows didn't do so well, or required more configuration)

**Comment by mariobuikhuizen** [comment]
Tip for multiple NVMe drives: make sure the platform has enough PCIe lanes.

I could not get Windows to scale beyond 2 drives with their soft-raid0 implementation, but I'm not a Windows expert. Linux scaled with all 7 PCIe 4.0 NVMe drives to 35 GB/s.

**Comment by Rakesh-E532238** [comment]
Awesome! I don't want to gamble with windows :) Sticking to Linux. 
@mariobuikhuizen thanks for the note. I will look out for PCle lanes

**Comment by Rakesh-E532238** [comment]
> Tip for multiple NVMe drives: make sure the platform has enough PCIe lanes.
> 
> I could not get Windows to scale beyond 2 drives with their soft-raid0 implementation, but I'm not a Windows expert. Linux scaled with all 7 PCIe 4.0 NVMe drives to 35 GB/s.

Do we need a PCIe lane for each NVMe drive?

**Comment by Rakesh-E532238** [comment]
I have a Dell Precision Tower 7910 with Dual Xeon E5-2690 which has 40 PCIe slots at 3.0.

Which NVMe drives suits best? Looking for either 1+1 (TB) or 2+2 TB. 

Please suggest

Thanks

**Comment by Rakesh-E532238** [comment]
I realized that those delays happen due to the read rate from your disk. The HDDs I used had 140-190MB/s. What is your datasets size and data read speed from disk?

This is how I calculate prior to any run nowadays. Time=(datasets size)/(read rate from disk)

**Comment by BustaRams** [comment]
Yeah I have about the same ~150ish MB/s, and the dataset is about 100Go.

**Comment by maartenbreddels** [comment]
Thanks for sharing this. I don't understand why your first cell takes 30 minutes. Could you split each line in a cell to see what is taking all the time.

100GB/150MB/s ~= 10 minutes, but Vaex should only touch the columns that are used in the filter, so it should basically be 1 column size / 150mb/s. Something seems wrong here (could be Vaex!).

**Comment by BustaRams** [comment]
Hi !

Here is the first cell broken down line by line 
![vaex3](https://user-images.githubusercontent.com/24378381/84389896-0d803800-abf7-11ea-8fd4-6f03b75a5c55.PNG)

> 100GB/150MB/s ~= 10 minutes, but Vaex should only touch the columns that are used in the filter, so it should basically be 1 column size / 150mb/s. Something seems wrong here (could be Vaex!).

Yeah that's what I thought, can't figure out where the problem comes from. Don't think it's vaex though as as it seems to work well for pretty much everyone else

**Comment by Rakesh-E532238** [comment]
@maartenbreddels I did test vaex against ~1 TB data and similar data split into 320GB and 640GB. The performance is more or less same when we calculate the statistics but fails when I try to calculate df.head() throwing a shape value error.

I have a question here! The data I mentioned above has all numericals. Now I have received a new dataset which has all the columns in String datatype. Can I draw the same performances with Vaex using strings compared to what I got using numericals? Any additional pre-processing that I should be aware of?
